### ISSUE TITLE
Disallow re-setting attributes

### ISSUE THREAD

**Issue by fornellas** [facebookincubator/TestSlide#156]
Currently from the DSL users can arbitrarily re-set context attributes:

```
@context
def top(context):
  @context.before
  def first_before(self):
    self.base = "base"
    self.derived = f"derived: {self.base}"

  @context.sub_context
  def inner(context):
    @context.before
    def second_before(self):
      self.base = "new base"
    @context.example
    def composition(self):
      self.assertEqual(self.derived, "derived: new base")  # Not True!
```

This is bad, as it prevents composition to work (eg: a sub context define a new value for an attribute).

We can change this, so that context attributes can be set **only once**.

This can probably be somewhat easily implemented by defining `_ContextData.__setattr__` with the "set once" rule.

**Comment by deathowl** [comment]
@fornellas 
I added a naive implementation of this on my local fork
```
diff --git a/testslide/__init__.py b/testslide/__init__.py
index e013bd7..8254e81 100644
--- a/testslide/__init__.py
+++ b/testslide/__init__.py
@@ -115,6 +115,15 @@ class _ContextData(object):
     def _all_memoizable_attributes(self):
         return self._context.all_context_data_memoizable_attributes

+    def __setattr__(self, name, value):
+        if self.__dict__.get(name) and self.__dict__[name] != value:
+            raise AttributeError(
+                f"Attribute: {name} is already set.\
+        Resetting attributes in sub-contexts is not allowed, as it breaks composition"
+            )
+        else:
+            super(_ContextData, self).__setattr__(name, value)
+
     def __getattr__(self, name):
         if name in self._all_methods.keys():

```

However, it breaks current testslide test suite, since we're resetting a couple of attributes in sub-contexts :
```
➜  TestSlide git:(disallow_resetting_attributes) ✗ make
COVERAGE ERASE
UNITTEST tests/cli_unittest.py
UNITTEST tests/dsl_unittest.py
UNITTEST tests/matchers_unittest.py
UNITTEST tests/testcase_unittest.py
TESTSLIDE tests/lib_testslide.py
TESTSLIDE tests/mock_async_callable_testslide.py
mock_async_callable()
  when target is a module
    and callable is a function
      can not mock
      can mock with flag
    and callable is an async function
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
  when target is a class
    and callable is a sync instance method
      can not mock
    and callable is a sync class method
      can not mock
      can mock with flag
    and callable is a sync static method
      can not mock
      can mock with flag
    and callable is a sync magic method
      can not mock
    and callable is an async instance method
      it is not allowed
    and callable is an async class method
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
    and callable is an async static method
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
    and callable is an async magic method
      it is not allowed
  an instance
    and callable is a sync instance method
      can not mock
      can mock with flag
    and callable is a sync class method
      can not mock
      can mock with flag
    and callable is a sync static method
      can not mock
      can mock with flag
    and callable is a sync magic method
      can not mock
      can mock with flag
    and callable is an async instance method
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
    and callable is an async class method
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
    and callable is an async static method
      default behavior
      .for_call()
      .to_raise(exception)
      .to_call_original()
      .and_assert_*
      .to_return_value(value)
        it returns value
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .to_return_values(value_list)
        it returns values
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
      .with_implementation(func)
        it calls mocked function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          raises NotACoroutine with non async function
      .with_wrapper(func)
        it calls function
        return value type
          passes with valid type
          with invalid return type
            raises TypeCheckError
        with sync function
          it raises NotACoroutine
    and callable is an async magic method
      default behavior: AttributeError: Attribute: call_args is already set.        Resetting attributes in sub-contexts is not allowed, as it breaks composition

Failures:

  1) mock_async_callable(), an instance, and callable is an async magic method: default behavior
    1) AttributeError: Attribute: call_args is already set.        Resetting attributes in sub-contexts is not allowed, as it breaks composition
      File "testslide/runner.py", line 600, in run
        self._run_example(example)
      File "testslide/runner.py", line 584, in _run_example
        _ExampleRunner(example, self.formatter).run()
      File "testslide/__init__.py", line 447, in run
        self._async_run_all_hooks_and_example(context_data)
      File "testslide/__init__.py", line 378, in _async_run_all_hooks_and_example
        asyncio.run(coro, debug=True)
      File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/asyncio/runners.py", line 43, in run
        return loop.run_until_complete(main)
      File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
        return future.result()
      File "testslide/__init__.py", line 289, in _real_async_run_all_hooks_and_example
        aggregated_exceptions.raise_correct_exception()
      File "testslide/__init__.py", line 207, in raise_correct_exception
        raise self.exceptions[0]
      File "testslide/__init__.py", line 191, in catch
        yield
      File "testslide/__init__.py", line 275, in _real_async_run_all_hooks_and_example
        before_code, context_data
      File "testslide/__init__.py", line 246, in _fail_if_not_coroutine_function
        return await func(*args, **kwargs)
      File "tests/mock_async_callable_testslide.py", line 562, in before
        self.call_args = ()
      File "testslide/__init__.py", line 122, in __setattr__
        Resetting attributes in sub-contexts is not allowed, as it breaks composition"
```

**Comment by fornellas** [comment]
Ahahaha, and the patch is exposing existing bugs at TestSlide's own codebase, lol.

The solution to **not** reset attributes, is to [memoize](https://testslide.readthedocs.io/en/master/testslide_dsl/context_attributes_and_functions/index.html#memoized-attributes) attributes using [composition](https://testslide.readthedocs.io/en/master/testslide_dsl/context_attributes_and_functions/index.html?highlight=composition#composition). We should probably mention this at the exception message.

Unfortunately, this is easier said than done, lol. I had a look, some should be easy to fix, but some... will be painful (bad existing test design...).

I can help with the hard bits, we can do this together if you want.

**Comment by deathowl** [comment]
Fixed by #237
